docs/source/expression.rst

```diff
@@ -0,0 +1,158 @@
+.. Copyright (c) 2016, Johan Mabille and Sylvain Corlay
+
+   Distributed under the terms of the BSD 3-Clause License.
+
+   The full license is in the file LICENSE, distributed with this software.
+
+Expressions and lazy evaluation
+===============================
+
+`xtensor` is more than an N-dimensional array library: it's an expression engine that allows numerical computation on any object implementing the expression interface.
+These objects can be in-memory containers such as ``xarray<T>`` and ``xtensor<T>``, but can also be backed by a database or a representation on the file system. This
+also enables creating adaptors as expressions for other data structures.
+
+Expressions
+-----------
+
+Assume ``x``, ``y`` and ``z`` are arrays of *broadcastable shapes* (we'll come back to that later), the return type of an expression such as ``x + y * sin(z)`` is **not an array**.
+It is an ``xexpression`` object offering the same interface as an N-dimensional array. You can store the result of such an expression and reuse it to build even more complex
+expressions:
+
+.. code::
+
+    auto f = x + y * sin(z);
+    auto f2 = w + 2 * cos(f);
+
+The expression engine avoids the evaluation of intermediate results and their storage in temporary arrays, so you can achieve the same performance as if you had written
+a simple loop. Assuming ``x``, ``y`` and ``z`` are one-dimensional arrays of length ``n``, 
+
+.. code::
+
+    xt::array<double> res = x + y * sin(z)
+   
+will produce quite the same assembly as the following loop:
+
+.. code::
+
+    xt::array<double> res(n);
+    for(size_t i = 0; i < n; ++i)
+    {
+        res(i) = x(i) + y(i) * sin(z(i));
+    }
+
+Lazy evaluation
+---------------
+
+An expression such as ``x + y * sin(z)`` does not hold the result. **Values are only computed upon access or when the expression is assigned to an xarray object**. This
+allows to operate symbolically on very large arrays and only compute the result for the indices of interest:
+
+.. code::
+
+    //Assume x and y are arrays each containing 1 000 000 objects
+    auto f = cos(x) + sin(y);
+
+    double first_res = f(1200);
+    double second_res = f(2500);
+    // Only two values have been computed
+
+That means if you use the same expression in two assign statements, the computation of the expression will be done twice. Depending on the complexity of the computation
+and the size of the data, it might be convenient to store the result of the expression in a temporary:
+
+.. code::
+
+    // Assume x and y are small arrays
+    xt::array<double> tmp = cos(x) + sin(y);
+    xt::array<double> res1 = tmp + 2 * x;
+    xt::array<double> res2 = tmp - 2 * x;
+
+Broadcasting
+------------
+
+The number of dimension of an ``xexpression`` and the sizes of these dimensions are described by the shape of the ``xexpression``, a container-like
```

Comment:
- dimension**s**
 - are provided by the `shape()` method, which returns a sequence of unsigned integers specifying the size of each dimension.